dcos/dcos-ui
# DCOS-34561: flatten the service tree to avoid extra lookups

This is an optimization to improve on the time it takes to render a page with many tasks. The improvement relies on the fact that lookin up the information about the tasks on a ServiceTree is rather expensive, and since it is performed once, by every task when we render a list of tasks, it does not scale for the number of tasks a big cluster have. The solution is to cache the ServiceTree as a lookup table, so you only pay the cost of looking the information once. It is a slow solution for very few number of tasks, but it pays of when you have a lot of them.

*How to test?:*
Just open some task pages, and the task nodes page and verify that the tasks with the data you need exist there. And is correct

If you care for more detail:.

Given we have **T** tasks to render:

**Before:**
We would perform 2 tree lookups for every task **T** on the ServiceTree. The ServiceTree has at most **T** tasks as well, and since it is usually not balanced (because some services might have a disproportional number of tasks), the worst case scenario for a lookup would be `O(T)`, making render the page performing **T** lookups of `O(T)` take `O(T^2)` worst case scenario. 

**With this fix:**
We fix that by flattening the tree into a lookup table once `O(T)`, and then performing as many lookups as we want in `O(1)*` time. Making render the full page cost the caching step **T** plus **T** lookups of constant cost with a total worse case of `O(T)`. We use extra memory proporcional to O(T) to do build the lookup table.

We could be even smarter and first filter the ServiceTree so it is cheap for us to answer: given a task, which service it belongs to.

**Checklist**
